pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/SchemaDefinition.java

```diff
@@ -76,4 +78,18 @@
      * @return the flag of supportSchemaVersioning
      */
     boolean getSupportSchemaVersioning();
+
+    /**
+     * Get a configured schema reader.
+     *
+     * @return optional containing configured schema reader or empty optional if none is configure
+     */
+    Optional<SchemaReader<T>> getSchemaReaderOpt();
```

Comment:
This interface is used for more than just JSON schemas, right? If so, it will be confusing for users to be able to set a reader and writer for any type of schema when they will only be used for JSON schemas and ignored otherwise. So we either need to support this broadly, or instead we can overload `JSONSchema.of()` for a more targeted approach.

Reply:
This interface can be used for other schemas as well, but both SchemaReader and SchemaWriter as well. I see what you mean, did not recognize the Producer and Consumer instantiators with Schema as parameter. With that in mind, yes for me it would be a beeter solution to overload SChema.of instead of extending this general purpose interface.

Reply:
Hmm, you'll probably want to overload `Schema.JSON(Class<T>)` and `Schema.JSON(SchemaDefinition)` as well as `JSONSchema.of()`. All of these are referenced in [the docs](https://pulsar.apache.org/docs/en/client-libraries-java/) so presumably they are meant for public use. (Though I would have guessed `JSONSchema` is not meant for public use since it's in the `impl` package. 🤷)